src/solc_0.7/catalyst/GemsCatalystsRegistry.sol

```diff
@@ -0,0 +1,154 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.7.1;
+pragma experimental ABIEncoderV2;
+
+import "./Gem.sol";
+import "./CatalystToken.sol";
+import "./AssetAttributesRegistry.sol";
+import "../common/BaseWithStorage/WithAdmin.sol";
+
+/// @notice Contract managing the Gems and Catalysts
+/// Each Gems and Catalys must be registered here.
+/// Each new Gem get assigned a new id (starting at 1)
+/// Each new Catalyst get assigned a new id (starting at 1)
+contract GemsCatalystsRegistry is WithAdmin {
+    function getAttributes(
+        uint16 catalystId,
+        uint256 assetId,
+        AssetAttributesRegistry.GemEvent[] calldata events
+    ) external view returns (uint32[] memory values) {
+        CatalystToken catalyst = getCatalyst(catalystId);
+        require(catalyst != CatalystToken(0), "CATALYST_DOES_NOT_EXIST");
+        return catalyst.getAttributes(assetId, events);
+    }
+
+    /// @notice Returns the maximum number of gems for a given catalyst
+    /// @param catalystId catalyst identifier
+    function getMaxGems(uint16 catalystId) external view returns (uint8) {
+        CatalystToken catalyst = getCatalyst(catalystId);
+        require(catalyst != CatalystToken(0), "CATALYST_DOES_NOT_EXIST");
+        return catalyst.getMaxGems();
+    }
+
+    /// @notice Burns one gem unit from each gem id on behalf of a beneficiary
+    /// @param from address of the beneficiary to burn on behalf of
+    /// @param gemIds list of gems to burn one gem from each
+    function burnDifferentGems(address from, uint16[] calldata gemIds) external {
+        for (uint256 i = 0; i < gemIds.length; i++) {
+            burnGem(from, gemIds[i], 1);
+        }
+    }
+
+    /// @notice Burns one catalyst unit from each catalyst id on behalf of a beneficiary
+    /// @param from address of the beneficiary to burn on behalf of
+    /// @param catalystIds list of catalysts to burn one catalyst from each
+    function burnDifferentCatalysts(address from, uint16[] calldata catalystIds) external {
+        for (uint256 i = 0; i < catalystIds.length; i++) {
+            burnCatalyst(from, catalystIds[i], 1);
```

Comment:
would be good to optimize the case where you burn 4 times the same gemID
that is what the previous broken implementation was trying to do